geom/twkb_parser.go

```diff
@@ -0,0 +1,525 @@
+package geom
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+)
+
+// Tiny Well Known Binary
+// See spec https://github.com/TWKB/Specification/blob/master/twkb.md
+
+// UnmarshalTWKB parses a Tiny Well Known Binary (TWKB), returning the
+// corresponding Geometry.
+func UnmarshalTWKB(twkb []byte, opts ...ConstructorOption) (Geometry, error) {
+	p := newTWKBParser(twkb, opts...)
+	geom, err := p.nextGeometry()
+	return geom, p.annotateError(err)
+}
+
+// TWKBParser holds all state information for interpreting TWKB buffers
+// including information such as the last reference point used in coord deltas.
+type TWKBParser struct {
+	twkb []byte
+	pos  int
+	opts []ConstructorOption
+
+	kind  int
+	ctype CoordinatesType
+
+	dimensions int
+	precX      int
+	precY      int
+	hasZ       bool
+	hasM       bool
+	precZ      int
+	precM      int
+	scalings   [twkbMaxDimensions]float64
+
+	hasBBox bool
+	hasSize bool
+	hasIDs  bool
+	hasExt  bool
+	isEmpty bool
+
+	bbox []int64
+
+	refpoint [twkbMaxDimensions]int64
+}
+
+func newTWKBParser(twkb []byte, opts ...ConstructorOption) TWKBParser {
+	return TWKBParser{
+		twkb:       twkb,
+		opts:       opts,
+		ctype:      DimXY,
+		dimensions: 2,
+	}
+}
+
+// Annotate an error message with the byte offset where it happened.
+func (p *TWKBParser) annotateError(err error) error {
+	if err == nil {
+		return nil
+	}
+	return fmt.Errorf("TWKB parsing error at byte %d: %s", p.pos, err.Error())
```

Comment:
The `%w` verb allows errors to be used with `errors.Is` and `errors.As`.

See: https://go.dev/blog/go1.13-errors

```suggestion
	return fmt.Errorf("TWKB parsing error at byte %d: %w", p.pos, err)
```

Reply:
I noticed a few more cases in this PR where we could use the `%w` verb as well.

Reply:
Nice, I'm using the `%w` verb now.